### PR TITLE
source-firestore: Implement Async Backfills (+ bump to v3)

### PR DESCRIPTION
**Pull request by willdonnelly**
**Description:**

This PR adds a new `backfillMode` property to resource bindings which can be one of:

- `none`: Skip preexisting data, just capture new documents and changes to existing documents.
- `sync`: The old behavior, where we ask Firestore for "all changes since Year 0" and just relay what comes back.
- `async`: The new behavior added by this PR, which streams new changes like `none` but then runs a backfill worker thread in parallel which fetches documents one chunk at a time from the database and emits those.

That's pretty much all done by the first commit in the sequence, what comes after that is:

- Some refactoring of the "generic gRPC capture testbench" test support code which makes it possible to verify massive and incremental captures.
- Some new "tests" in `datasynth_test.go` which aren't so much automated tests as they are tools for a human to semi-manually verify that we can successfully and reliably capture massive datasets, by generating such a synthetic dataset and then capturing it.
- Making `async` the default suggested by discovery, modifying the preexisting test suite to exercise the async backfill mode, and fixing a couple of edge case bugs that this turned up.
- Various cleanup work, documentation, etc.

**Workflow steps:**

Preexisting users of `source-firestore` (all one or two of them) will need to add `"backfillMode": "sync"` to all preexisting capture bindings when they update to use the `:v3` connector, and things will keep working pretty much unchanged. If they manually add new capture bindings in the future, it is recommended to use `"async"`.

New users will get `"backfillMode": "async"` by default if they use discovery.

**Documentation links affected:**

https://docs.estuary.dev/reference/Connectors/capture-connectors/google-firestore/ will need some changes to mention the `backfillMode` property and what the different settings mean. There's also some stuff I wrote in the README about the Firestore data model and how we map that to collections that might be worth adding.

- - -
This change is [Reviewable](https://reviewable.io/reviews/estuary/connectors/395)

### CODE REVIEW

File: source-firestore/pull.go [hunk 1]
```diff
@@ -115,27 +122,91 @@ type captureState struct {
...
+type backfillState struct {
+	Cursor string    // The last document backfilled
+	MTime  time.Time // The UpdateTime of that document when we backfilled it
+}
+
+func (s *backfillState) Equal(x *backfillState) bool {
+	if s == nil {
+		return x == nil
+	} else if x == nil {
+		return false
```
[the remark this answers is not in the document]
Comment (willdonnelly) [reply source-firestore/pull.go:138]: In order to deal properly with various edge cases around adding/removing bindings, capture state is tracked per resource path, but the Firestore APIs for streaming and querying data all operate in terms of collection groups based on the final path element. So during startup there's a bit of logic (see line 333) which amounts to:
1. Resume streaming from the oldest resume timestamp associated with a particular collection group (and then timestamp-based filtering on a per-resource basis will prevent duplicate already-processed events).
2. Also perform at most one backfill from a given collection group, resuming from the persisted state if it's the same for all resources in that group and resetting back to the start if there's a mismatch.

The `Equal()` helper method helps to make that logic slightly clearer, since `nil` is a valid backfill state meaning "no backfill is ongoing or necessary right now".

**Review by williamhbaker — APPROVED** [review]

File: source-firestore/pull.go [hunk 2]
```diff
@@ -289,40 +399,157 @@ func (c *capture) Run(ctx context.Context) error {
...
+		return nil
+	} else if resumeState.Cursor == "" {
+		// If the cursor path is empty then we just leave the cursor document pointer nil
+		logEntry.Info("starting async backfill")
+	} else if resumeDocument, err := client.Doc(resumeState.Cursor).Get(ctx); err != nil {
+		// If we fail to fetch the resume document, we clear the relevant backfill cursor and error out.
+		// This will cause the backfill to restart from the beginning after the capture gets restarted,
+		// and in the meantime it will show up as an error in the UI in case there's a persistent issue.
+		if checkpointJSON, err := c.State.UpdateBackfillState(collectionID, &backfillState{}); err != nil {
+			return err
```
[the remark this answers is not in the document]
Comment (willdonnelly) [reply source-firestore/pull.go:446]: Yeah, if `UpdateBackfillState` returns an error then something has gone very wrong, and "figure out a merge patch for the collection group being worked on" is like 50% of its job already, so I don't think there's a good fallback from that.

File: source-firestore/pull.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mdibaiee) [reply source-firestore/pull.go:138]: Okay, I understand that it has value, but the naming is confusing. I would strongly expect `x.Equal(nil)` and `nil.Equal(x)` to be the same. I see that this is only used in one place, do you think we could rename this to something more suitable for its use?

File: source-firestore/pull.go [hunk 3]
```diff
@@ -289,40 +399,157 @@ func (c *capture) Run(ctx context.Context) error {
...
+
+		for _, doc := range docs {
+			logEntry.WithField("doc", doc.Ref.Path).Trace("got document")
+
+			// We update the cursor before checking whether this document is being
+			// backfilled. This does, unfortunately, mean that it's possible for changes
+			// to a document which *isn't being captured* could break the ongoing backfill
+			// resume behavior, but that's just how Firestore collection group queries
+			// work.
+			cursor = doc
```
[the remark this answers is not in the document]
Comment (willdonnelly) [reply source-firestore/pull.go:496]: The cursor is used for every chunk query of 4,096 documents, so if we don't update it to a non-captured document at all and there are more than 4,096 such documents in a row we'll just spin forever and not make forward progress. But I think I see what you're getting at, we could decouple the cursor used during a backfill a bit from what gets persisted to the state checkpoint, so the persisted resume cursor will only ever be a captured document.

However this would introduce a potential failure mode where the collection group queries start with a massive dataset that isn't supposed to be captured, we start working our way through that, and then get restarted repeatedly and basically we keep causing a ton of read traffic without actually making forward progress (until the connector is left running long enough that it can complete all of that and reach a captured document and update the state checkpoint to that). I'm not sure if this is preferable, but either way I think we can adjust the behavior in the future if it proves necessary.

File: source-firestore/pull.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mdibaiee) [reply source-firestore/pull.go:138]: ah I see, you are right this is commutative, I got confused by the if conditions 👍🏽

File: source-firestore/pull.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mdibaiee) [reply source-firestore/pull.go:496]: That sounds reasonable, I wanted to make sure I understand the implication of this 👍🏽
